### DISCUSSION TITLE
SPICE netlist speedup

### DISCUSSION THREAD

**Discussion by Vinc0110** [scikit-rf/scikit-rf#1225] [Ideas]
Based on the discussion with @fnordpole (https://github.com/scikit-rf/scikit-rf/discussions/1172) I had some ideas to further streamline and maybe accelerate the SPICE simulations of the synthesized netlists after vector fitting.

Looking into the numeric details of circuit simulations with SPICE, which are based on (modified) nodal analysis (MNA) [1], [2], one can get an idea of the computational workload caused by the different circuit elements. More specifically, I was wondering if controlled sources are computationally expensive and if there are any differences between the different types of controlled sources (VCCS, CCCS, VCVS, CCVS).

Apparently, only current sources and VCCS are supported by plain nodal analysis, all other sources require a modification (MNA), for which the conductance matrix is extended. Of course, it takes more time to solve for the solution vector when the matrix is large.

Comparing the "stamp" in the coefficient matrix of different controlled sources (see [2], page 22-24 and pages 34-37), VCCS has the smallest stamp, which is as small as the stamp of a resistor:
- R: 2x2 stamp
- VCCS: 2x2 stamp
- V: 3x3 stamp
- CCCS: 5x5 stamp (controlling current can be reused in case of multiple CCCS)
- CCVS: 6x6 stamp (controlling current can be reused in case of multiple CCVS)
- VCVS: 5x5 stamp

So, I made two test networks similar to the existing synthesized circuits currently used in the VectorFitting module. One with many admittances in parallel driven by controlled voltage sources (VCVS and CCVS), and one with many impedances in series driven by controlled current sources (VCCS and CCCS). Also for the port interface, I compared Thévenin sources (CCVS with resistor in series) and Norton sources (VCCS with resistor in parallel).

I'll share details of the circuits later, but here are some preliminary results:
- 100 admittances take **0.244 s**, 100 impedances take **0.019 s**
- 1000 admittances take **346.1s**, 1000 impedances take **0.182 s**

In VectorFitting, we are currently using admittances with controlled voltage sources, so it seems there is a massive opportunity for speed improvements if we switch to impedances with controlled current sources. However, the equivalent impedance networks for complex-conjugate pairs include CSVS, whereas the currently used equivalent admittance networks use VCCS. 

I should probably finish a full implementation of the netlist synthesizer and then make a fair comparison against the current implementation. Hopefully, I can find some more time for this in the next days...

[1] http://emlab.illinois.edu/ece546/Lect_16.pdf
[2] https://aice.sjtu.edu.cn/msda/data/courseware/SPICE/lect07_element_stamping.pdf

**Comment by fnordpole** [comment]
Sounds like a promising idea to convert the hole thing to impedance-based instead of admittance-based!

According to this PDF, both spice and also Xyce are based on MNA, so we could get this speed up probably across all those simulators: https://xyce.sandia.gov/files/xyce/Xyce_Math_Formulation.pdf

The current implementation uses a VCVS+CCVS pair for the incident wave (a), a VCVS for negative incident wave (-a) per port and one CCVS for the port network and one VCCS per admittance of a complex conjugate pole/residue pair.

So we have 
n_ports * 4 controlled voltage sources
n_poles_complex * n_ports * n_ports * 1 controlled current-source

In a potential new implementation  based on impedances instead of admittances, everything will be approximately the opposite:
n_ports * 4 controlled current sources
n_poles_complex * n_ports * n_ports * 1 controlled voltage sources.

Because the second term goes with n_ports quadratically, the new implementation would probably only be faster for very small n_ports. So I guess, we would need to find a solution to get rid of the CCVS in the impedance realization!

There is indeed in
"SPICE Equivalent Circuits of Frequency-Domain Responses", Giulio Antonini, 
a second realization without a VCCS for a complex conjugate pole pair that has no controlled source:

![antonini-second](https://github.com/user-attachments/assets/b366f573-69eb-4957-99b8-ac192bc911a2)

Maybe something similar but for an impedance instead of an admittance could be used to get rid of the CCVS. However, I don't know about potential drawbacks of this non-active-source-realizations. 

Anyway, it's a good idea and I am looking forward to the results of it! I did not dive so deep into the simulators to think about these type of optimizations.

**Comment by Vinc0110** [comment]
Antonini's paper is great. It's providing equivalent active and passive networks for both admittance and impedance representations. The passive equivalent impedance network that you showed above only works in some cases. For some pole-residue pairs, the conductances G1 and G2 would need to have negative values. In other words, gain is required, which is often the case for highly resonant poles. This makes absolute sense, because we would also like to represent active S-parameter responses. While Antonini does not mention this problem at all in his paper, he does provide another circuit with a series CCVS.

As discusses earlier, CCVS are the worst case in terms of computational efficiency in modified nodal analysis. VCCS would be much better. They basically come for free. So, I came up with the following circuit, which is able to generate negative conductances with the help of a VCCS in parallel to the conductances. This seems to work very well. Please also check PR #1228.

![circuit_equivalent_impedance_cc](https://github.com/user-attachments/assets/f9c52be8-8280-408c-91e1-50570f9faa71)
